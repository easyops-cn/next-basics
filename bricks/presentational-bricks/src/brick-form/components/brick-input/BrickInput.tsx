import React from "react";
import { Input } from "antd";
import { WrapperFormItem } from "../wrapper-form-item/WrapperFormItem";
import { InputProps } from "antd/lib/input";
import { handleChange } from "../../processor";
export enum LegacyBrickInputType {
  Input = "Input",
  Password = "Password",
  TextArea = "TextArea"
}

export interface LegacyBrickInputProps {
  configProps?: InputProps;
  type?: LegacyBrickInputType;
  onChange: Function;
  value?: string;
}

export function LegacyBrickInput({
  configProps,
  onChange,
  value,
  type = LegacyBrickInputType.Input
}: LegacyBrickInputProps): React.ReactElement {
  const renderComp = (Component: any) => (
    <Component
      {...configProps}
      value={value}
      onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
        handleChange(e, onChange)
      }
    />
  );

  const map = {
    [LegacyBrickInputType.Input]: Input,
    [LegacyBrickInputType.Password]: Input.Password,
    [LegacyBrickInputType.TextArea]: Input.TextArea
  };

  return renderComp(map[type]);
}

export const BrickInput = WrapperFormItem(LegacyBrickInput);
