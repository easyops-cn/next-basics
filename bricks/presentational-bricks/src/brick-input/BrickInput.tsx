import React, { useState, useEffect, useMemo } from "react";
import { SearchOutlined } from "@ant-design/icons";
import { Input } from "antd";
import style from "./BrickInput.module.css";
import { debounce } from "lodash";

export interface BrickInputProps {
  defaultValue?: string;
  value?: string;
  handleValueEmit: any;
  handleValueChange: any;
  placeholder?: string;
  trigger?: "change" | "enter";
  debounceTime: number;
}

export function BrickInput(props: BrickInputProps): React.ReactElement {
  const [inputVal, setInputVal] = useState(props.defaultValue);

  const debounceValueEmit = useMemo(
    () =>
      debounce((value: string): void => {
        props.handleValueEmit(value);
      }, props.debounceTime),
    [props.debounceTime]
  );

  useEffect(() => {
    props.handleValueChange && props.handleValueChange(props.value);
    setInputVal(props.value);
  }, [props.value]);

  const handleChange = ({ target: { value } }: any) => {
    setInputVal(value);
    if (props.trigger !== "enter") {
      debounceValueEmit(value);
    }
    props.handleValueChange && props.handleValueChange(value);
  };

  const handleValEmit = () => {
    if (props.trigger === "enter") {
      props.handleValueEmit(inputVal);
    }
  };

  return (
    <div style={{ position: "relative" }}>
      <Input
        placeholder={props.placeholder}
        className={style.searchInput}
        value={inputVal}
        onChange={handleChange}
        onPressEnter={handleValEmit}
      ></Input>
      <SearchOutlined className={style.searchIcon} />
    </div>
  );
}
