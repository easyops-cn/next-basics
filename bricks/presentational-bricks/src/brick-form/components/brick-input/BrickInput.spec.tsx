import React from "react";
import { shallow } from "enzyme";
import { LegacyBrickInput, LegacyBrickInputType } from "./BrickInput";

describe("BrickInput", () => {
  const onChange = jest.fn();
  it("should work", () => {
    const wrapper = shallow(<LegacyBrickInput onChange={onChange} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should trigger change event", () => {
    const wrapper = shallow(
      <LegacyBrickInput
        onChange={onChange}
        type={LegacyBrickInputType.TextArea}
      />
    );
    wrapper.simulate("change", { target: { value: "x" } });
    expect(onChange).toHaveBeenCalled();
  });
});
