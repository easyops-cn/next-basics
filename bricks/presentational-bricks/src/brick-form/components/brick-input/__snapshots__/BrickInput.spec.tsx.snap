// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickInput should work 1`] = `
<Input
  onChange={[Function]}
  type="text"
/>
`;
